<a href="https://colab.research.google.com/github/bruno-ba/python-lab/blob/main/exercicios_python_brasil.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Desafio Devpro

Nesse notebook vamos resolver vários exercícios da [lista da Python Brasil](https://wiki.python.org.br/ListaDeExercicios)

# [Estrutura Sequencial](https://wiki.python.org.br/EstruturaSequencial)

In [ ]:
import math

Exercício 1

In [ ]:
print('Olá, Mundo')

Olá, Mundo


In [ ]:
type('Olá, Mundo')

str

Exercício 2

In [ ]:
num  = input('Digite um número: ')
print( f'O número informado foi {num}')

Digite um número: 852
O número informado foi 852


Exercício 3

In [ ]:
num_1  = int(input('Digite um número: '))
num_2  = int(input('Digite outro número: '))
soma = num_1+num_2
print(f'A soma dos dois números é {soma}')

Digite um número: 12
Digite outro número: 5
A soma dos dois números é 17


Exercício 4

Faça um Programa que peça as 4 notas bimestrais e mostre a média. 

In [ ]:
num_1 = float(input('Digite a primeira nota: '))
num_2 = float(input('Digite a segunda nota: '))
num_3 = float(input('Digite a terceira nota: '))
num_4 = float(input('Digite a quarta nota: '))

media = (num_1 + num_2 + num_3 + num_4) / 4

print(f'A média das 4 notas bimestrais é: {media}')

Digite a primeira nota: 8.64
Digite a segunda nota: 9.59
Digite a terceira nota: 6.98
Digite a quarta nota: 5.89
A média das 4 notas bimestrais é: 7.775


Exercício 5

Faça um Programa que converta metros para centímetros. 

In [ ]:
num_1 = float(input('Digite o valor em metros: '))
num_2 = num_1 * 100
print(f'O valor em metros: {num_1}m, valor em centímetros: {num_2}cm')

Digite o valor em metros: 5
O valor em metros: 5.0m, valor em centímetros: 500.0cm


Exercício 6

Faça um Programa que peça o raio de um círculo, calcule e mostre sua área. 

In [ ]:
num_1 = float(input('Informe o raio do círculo: '))
area =  math.pi * num_1 ** 2
print(f'A área do círculo π * ({num_1})² = {area}')


Informe o raio do círculo: 18
A área do círculo π * (18.0)² = 1017.8760197630929


### Exercício 7

Faça um Programa que calcule a área de um quadrado, em seguida mostre o dobro desta área para o usuário. 


In [ ]:
num_1 = float(input('Digite o lado do quadrado: '))
num_2 = num_1 ** 2 * 2
print(f'O dobro da área do quadrado ({num_1})² x 2 = {num_2}')

Digite o lado do quadrado: 10
O dobro da área do quadrado (10.0)² x 2 = 200.0


Exercício 8

Faça um Programa que pergunte quanto você ganha por hora e o número de horas trabalhadas no mês. Calcule e mostre o total do seu salário no referido mês. 

In [ ]:
num_1 = float(input('Quanto você ganha por hora? '))
num_2 = float(input('Qual o número de horas trabalhadas no mês? '))
num_3 = num_1 * num_2
print(f'O seu salário será de R$ {num_3}')

Quanto você ganha por hora? 20
Qual o número de horas trabalhadas no mês? 34.6
O seu salário será de R$ 692.0


Exercício 9

Faça um Programa que peça a temperatura em graus Fahrenheit, transforme e mostre a temperatura em graus Celsius.

    C = 5 * ((F-32) / 9). 

In [ ]:
num_1 = float(input('Digite a temperatura em graus Fahrenheit: '))
num_2 = 5 * ((num_1 - 32) / 9)
print(f'Graus Celsius 5 x (({num_1} - 32) ÷ 9) = {num_2:.2f} °C')

Digite a temperatura em graus Fahrenheit: 9
Graus Celsius 5 x ((9.0 - 32) ÷ 9) = -12.78 °C


Exercício 10

Faça um Programa que peça a temperatura em graus Celsius, transforme e mostre em graus Fahrenheit.
(x °C × 9/5) + 32 = 32 °F


In [ ]:
num_1 = float(input('Digite a temperatura em graus Celsius: '))
r_1 = num_1 * 9 / 5 + 32
print(f'({num_1}°C × 9 ÷ 5) + 32 = {r_1} °F')

Digite a temperatura em graus Celsius: 56.3
(56.3°C × 9 ÷ 5) + 32 = 133.34 °F


Exercício 11

Faça um Programa que peça 2 números inteiros e um número real. Calcule e mostre:

    o produto do dobro do primeiro com metade do segundo .
    a soma do triplo do primeiro com o terceiro.
    o terceiro elevado ao cubo. 

In [ ]:
num_1 = int(input('Digite o primeiro número inteiro: '))
num_2 = int(input('Digite o segundo número inteiro: '))
num_3 = float(input('Digite um número '))

r_1 = (2 * num_1) * (num_2 / 2)
r_2 = (3 * num_1) + num_3
r_3 = num_3 ** 3

print(f'{r_1, r_2, r_3}')

Digite o primeiro número inteiro: 3
Digite o segundo número inteiro: 9
Digite um número 5.6
(27.0, 14.6, 175.61599999999996)


Exercício 12

Tendo como dados de entrada a altura de uma pessoa, construa um algoritmo que calcule seu peso ideal, usando a seguinte fórmula: (72.7*altura) - 58 

In [ ]:
num_1 = float(input('Digite a sua altura: '))
r_1 = 72.7 * num_1 - 58
print(f'Peso ideal 72,7 x {num_1} - 58 = {r_1:.2f}')

Digite a sua altura: 1.82
Peso ideal 72,7 x 1.82 - 58 = 74.31


Exercício 13

Tendo como dado de entrada a altura (h) de uma pessoa, construa um algoritmo que calcule seu peso ideal, utilizando as seguintes fórmulas:

    Para homens: (72.7*h) - 58
    Para mulheres: (62.1*h) - 44.7 
    

In [ ]:
num_1 = float(input('Digite a altura: '))
r_1 = 72.7 * num_1 - 58
r_2 = 62.1 * num_1 - 44.7
print(f'''PESO IDEAL
Para homens: (72.7 x {num_1}) - 58 = {r_1:.2f}
Para mulheres: (62.1 x {num_1}) - 44.7 = {r_2:.2f}
''')

Digite a altura: 1.8
PESO IDEAL
Para homens: (72.7 x 1.8) - 58 = 72.86
Para mulheres: (62.1 x 1.8) - 44.7 = 67.08



Exercício 14

João Papo-de-Pescador, homem de bem, comprou um microcomputador para controlar o rendimento diário de seu trabalho. Toda vez que ele traz um peso de peixes maior que o estabelecido pelo regulamento de pesca do estado de São Paulo (50 quilos) deve pagar uma multa de R$ 4,00 por quilo excedente. João precisa que você faça um programa que leia a variável peso (peso de peixes) e calcule o excesso. Gravar na variável excesso a quantidade de quilos além do limite e na variável multa o valor da multa que João deverá pagar. Imprima os dados do programa com as mensagens adequadas. 

In [ ]:
peso_regulamento_sp = 50.00
multa_quilo_excedente = 4.00
peso_de_peixes = float(input('Digite o peso dos peixes: '))
excesso_calc = peso_de_peixes - peso_regulamento_sp
if excesso_calc <= 0:
  excesso_calc = 0
multa_calc = multa_quilo_excedente * excesso_calc
print(f'''"Cáculo de multa(R$) sobre excesso de peso(Kg) de peixes."
peso regulamentado: 50kg
multa sobre peso excedente: R$4.00

peso: {peso_de_peixes:.2f}Kg
peso em excesso: {excesso_calc:.2f}Kg
multa sobre excesso: R${multa_calc:.2f}

''')

Digite o peso dos peixes: 69.56
"Cáculo de multa(R$) sobre excesso de peso(Kg) de peixes."
peso regulamentado: 50kg
multa sobre peso excedente: R$4.00

peso: 69.56Kg
peso em excesso: 19.56Kg
multa sobre excesso: R$78.24




Exercício 15

Faça um Programa que pergunte quanto você ganha por hora e o número de horas trabalhadas no mês. Calcule e mostre o total do seu salário no referido mês, sabendo-se que são descontados 11% para o Imposto de Renda, 8% para o INSS e 5% para o sindicato, faça um programa que nos dê:

    salário bruto.
    quanto pagou ao INSS.
    quanto pagou ao sindicato.
    o salário líquido.
    calcule os descontos e o salário líquido, conforme a tabela abaixo:

    + Salário Bruto : R$
    - IR (11%) : R$
    - INSS (8%) : R$
    - Sindicato ( 5%) : R$
    = Salário Liquido : R$

    Obs.: Salário Bruto - Descontos = Salário Líquido. 
    

In [ ]:
val_sal_hr = float(input('Digite quanto você ganha por hora R$ '))
val_hr_trab_mes =  float(input('Digite o número de horas trabalhadas no mês. '))
calc_sal_bruto = val_sal_hr * val_hr_trab_mes
calc_inss = 0.08 * calc_sal_bruto
calc_ir = 0.11 * calc_sal_bruto
calc_sindicato = 0.05 * calc_sal_bruto
calc_descontos = calc_inss + calc_ir + calc_sindicato
calc_sal_liquido = calc_sal_bruto - calc_descontos

print(f'''
CÁCULO SALÁRIO
+ Salário Bruto: R${calc_sal_bruto:.2f}
- IR (11%): R${calc_ir:.2f}
- INSS (8%): R${calc_inss:.2f}
- Sindicato (5%): R${calc_sindicato:.2f}
= Salário Líquido: R${calc_sal_liquido:.2f}
''')


Digite quanto você ganha por hora R$ 38.65
Digite o número de horas trabalhadas no mês. 44.20

CÁCULO SALÁRIO
+ Salário Bruto: R$1708.33
- IR (11%): R$187.92
- INSS (8%): R$136.67
- Sindicato (5%): R$85.42
= Salário Líquido: R$1298.33



Exercício 16

Faça um programa para uma loja de tintas. O programa deverá pedir o tamanho em metros quadrados da área a ser pintada. Considere que a cobertura da tinta é de 1 litro para cada 3 metros quadrados e que a tinta é vendida em latas de 18 litros, que custam R$ 80,00. Informe ao usuário a quantidades de latas de tinta a serem compradas e o preço total. 

In [ ]:
VOLUME_LATA = 18.00
PRECO_LATA = 80.00
RENDIMENTO_TINTA = 3

area_pintura = float(input('Digite o tamanho da área a ser pintada(m²) '))
litros_tinta_necessario = area_pintura / RENDIMENTO_TINTA
latas_tinta_necessarias = litros_tinta_necessario / VOLUME_LATA 

if latas_tinta_necessarias % 1 != 0:
  latas_tinta_necessarias = latas_tinta_necessarias // 1 + 1

preco_latas_tinta_necessarias = latas_tinta_necessarias * PRECO_LATA

print(f'''
"Cáculo latas de tintas"
área a ser pintada = {area_pintura:.2f} m²
cobertura de tinta = {litros_tinta_necessario:.2f} L 
total de latas = {latas_tinta_necessarias} und
preço total = R${preco_latas_tinta_necessarias:.2f}
''')

Digite o tamanho da área a ser pintada(m²) 120

"Cáculo latas de tintas"
área a ser pintada = 120.00 m²
cobertura de tinta = 40.00 L 
total de latas = 3.0 und
preço total = R$240.00



Faça um Programa para uma loja de tintas. O programa deverá pedir o tamanho em metros quadrados da área a ser pintada. Considere que a cobertura da tinta é de 1 litro para cada 6 metros quadrados e que a tinta é vendida em latas de 18 litros, que custam  R\$80,00 ou em galões de 3,6 litros, que custam  R\$25,00.

Informe ao usuário as quantidades de tinta a serem compradas e os respectivos preços em 3 situações:

    comprar apenas latas de 18 litros;
    comprar apenas galões de 3,6 litros;
    misturar latas e galões, de forma que o desperdício de tinta seja menor. Acrescente 10% de folga e sempre arredonde os valores para cima, isto é, considere latas cheias. 

In [ ]:
VOLUME_LATA = 18.00
PRECO_LATA = 80.00
VOLUME_GALAO = 3.6
PRECO_GALAO = 25.00
RENDIMENTO_TINTA = 6
RAZAO_LATA_GALAO = VOLUME_LATA / VOLUME_GALAO

area_pintura = float(input('Digite o tamanho da área a ser pintada(m²) '))
area_com_folga = area_pintura * 1.1
litros_tinta_necessario = area_com_folga / RENDIMENTO_TINTA


apenas_latas_qtd = litros_tinta_necessario / VOLUME_LATA
apenas_galoes_qtd = litros_tinta_necessario / VOLUME_GALAO

apenas_latas_fracao =  apenas_latas_qtd % 1
apenas_galao_fracao = apenas_galoes_qtd % 1

if apenas_latas_fracao != 0:
  apenas_latas_qtd = apenas_latas_qtd // 1 + 1

if apenas_galao_fracao != 0:
  apenas_galoes_qtd = apenas_galoes_qtd // 1 + 1

apenas_latas_preco = apenas_latas_qtd * PRECO_LATA
apenas_galao_preco = apenas_galoes_qtd * PRECO_GALAO

lg_latas_qtd = litros_tinta_necessario / VOLUME_LATA
lg_latas_fracao = lg_latas_qtd % 1

if lg_latas_fracao == 0:
  lg_galoes_qtd =0

else:
  lg_latas_qtd = lg_latas_qtd // 1
  lg_galoes_qtd = lg_latas_fracao * RAZAO_LATA_GALAO
  lg_galoes_fracao =  lg_galoes_qtd % 1
  if lg_galoes_fracao != 0:
    lg_galoes_qtd = lg_galoes_qtd // 1 + 1

lg_latas_preco = lg_latas_qtd * PRECO_LATA
lg_galoes_preco = lg_galoes_qtd * PRECO_GALAO
lg_preco_total = lg_latas_preco + lg_galoes_preco

print(f'''
área a ser pintada = {area_pintura:.2f} m²
área com folga (10%) = {area_com_folga:.2f}m²
litros de tinta necessários = {litros_tinta_necessario:.2f} L

APENAS LATAS
quantidade de latas = {apenas_latas_qtd} und
preço = R${apenas_latas_preco:.2f}

APENAS GALÕES
quantidade de galões = {apenas_galoes_qtd} und
preço = R${apenas_galao_preco:.2f}

LATAS E GALÕES COM ECONOMIA DE TINTA
quantidade de latas = {lg_latas_qtd} und
quantidade de galões = {lg_galoes_qtd} und

preço latas = R${lg_latas_preco:.2f}
+
preço galões = R${lg_galoes_preco:.2f}
=
preço total = R${lg_preco_total}
''')



Digite o tamanho da área a ser pintada(m²) 200

área a ser pintada = 200.00 m²
área com folga (10%) = 220.00m²
litros de tinta necessários = 36.67 L

APENAS LATAS
quantidade de latas = 3.0 und
preço = R$240.00

APENAS GALÕES
quantidade de galões = 11.0 und
preço = R$275.00

LATAS E GALÕES COM ECONOMIA DE TINTA
quantidade de latas = 2.0 und
quantidade de galões = 1.0 und

preço latas = R$160.00
+
preço galões = R$25.00
=
preço total = R$185.0



In [ ]:
print(lg_latas_fracao)

0.55


# [Estrutura de Decisão](https://wiki.python.org.br/EstruturaDeDecisao)

Exercício 1

Faça um Programa que peça dois números e imprima o maior deles. 

In [ ]:
num_1 = float(input('Digite o primeiro número '))
num_2 = float(input('Digite o segundo número '))

maior_num = max(num_1, num_2)

print(f'O maior número é {maior_num}')

Exercício 2

Faça um Programa que peça um valor e mostre na tela se o valor é positivo ou negativo. 

In [ ]:
num_1 = float(input('Digite um número '))
if num_1 < 0:
  print(f'{num_1} é um número negativo')
else:
  print(f'{num_1} é um número positivo')

Exercício 3

Faça um Programa que verifique se uma letra digitada é "F" ou "M". Conforme a letra escrever: F - Feminino, M - Masculino, Sexo Inválido. 

In [ ]:
sexo = input('Digite uma letra: ')
if sexo in ['F', 'f']:
  print(f'{sexo} - Feminino')
elif sexo in ['M', 'm']:
  print(f'{sexo} - Masculino')
else:
  print(f'Sexo Inválido')


Exercício 4

Faça um Programa que verifique se uma letra digitada é vogal ou consoante.

In [ ]:
vogais = ['a','e','i','o','u']
letra_digitada = (input('Digite uma letra: ')).lower()
if letra_digitada in vogais:
  print(f'A letra digitada é uma vogal')
else:
  print(f'A letra digitada é uma consoante')


Exercício 5

Faça um programa para a leitura de duas notas parciais de um aluno. O programa deve calcular a média alcançada por aluno e apresentar:

    A mensagem "Aprovado", se a média alcançada for maior ou igual a sete;
    A mensagem "Reprovado", se a média for menor do que sete;
    A mensagem "Aprovado com Distinção", se a média for igual a dez. 

In [ ]:
nota_1 = float(input('Digite a primeira nota: '))
nota_2 = float(input('Digite a segunda nota: '))
media = (nota_1 + nota_2) / 2
if media == 10:
  print(f'"Aprovado com Distinção"')
elif media >= 7:
  print(f'"Aprovado"')
else:
  print(f'"Reprovado1"')

Exercício 6

Faça um Programa que leia três números e mostre o maior deles. 

In [ ]:
num_1 = float(input('Digite o primeiro número: '))
num_2 = float(input('Digite o segundo número: '))
num_3 = float(input('Digite o terceiro número: '))

maior = num_1
if num_2 > maior :
  maior = num_2

if num_3 > maior:
  maior = num_3

print(f'O maior número digitado foi {maior}')

Exercício 7

Faça um Programa que leia três números e mostre o maior e o menor deles. 

In [ ]:
num_1 = float(input('Digite o primeiro número: '))
num_2 = float(input('Digite o segundo número: '))
num_3 = float(input('Digite o terceiro número: '))

maior = num_1
menor = num_1

if num_2 > maior :
  maior = num_2

if num_3 > maior:
  maior = num_3

if num_2 < menor :
  menor = num_2

if num_3 < menor:
  menor = num_3

print(f'O maior número digitado foi {maior}')
print(f'O maior número digitado foi {menor}')

Exercício 8

Faça um programa que pergunte o preço de três produtos e informe qual produto você deve comprar, sabendo que a decisão é sempre pelo mais barato. 

In [ ]:
num_1 = float(input('Digite o preço do primeiro produto: '))
num_2 = float(input('Digite o preço do segundo produto: '))
num_3 = float(input('Digite o preço do terceiro produto: '))

produto = 'primeiro'
menor_preco = num_1
if num_2 < menor_preco:
   menor_preco = num_2
   produto='segundo'

if num_3 < menor_preco:
   menor_preco = num_3
   produto='terceiro'

print(f'Você deverá comprar o {produto} produto custando: R${menor_preco:.2f}')

Exercício 9

Faça um Programa que leia três números e mostre-os em ordem decrescente. 

In [ ]:
num_1 = float(input('Digite o primeiro número:  '))
num_2 = float(input('Digite o segundo número: '))
num_3 = float(input('Digite o terceiro número:  '))

if num_2 >= num_1:
  num_1, num_2 = num_2, num_1

if num_3 >= num_1:
  num_1, num_3 = num_3, num_1

if num_3 >= num_2:
  num_2, num_3 = num_3, num_2

print(f'{num_1}, {num_2}, {num_3} ')

Exercício 10

Faça um Programa que pergunte em que turno você estuda. Peça para digitar M-matutino ou V-Vespertino ou N- Noturno. Imprima a mensagem "Bom Dia!", "Boa Tarde!" ou "Boa Noite!" ou "Valor Inválido!", conforme o caso. 

In [ ]:
turno = (input('Digite o turno que você estuda: ')).lower()

if turno == 'm':
  print(f'Bom Dia!')

elif turno == 'v':
  print(f'Boa Tarde!')

elif turno == 'n':
  print(f'Boa Noite!')

else:
  print(f'Valor Inválido')

Exercício 11

As Organizações Tabajara resolveram dar um aumento de salário aos seus colaboradores e lhe contraram para desenvolver o programa que calculará os reajustes.

    Faça um programa que recebe o salário de um colaborador e o reajuste segundo o seguinte critério, baseado no salário atual:
    salários até R$ 280,00 (incluindo) : aumento de 20%
    salários entre R$ 280,00 e R$ 700,00 : aumento de 15%
    salários entre R$ 700,00 e R$ 1500,00 : aumento de 10%
    salários de R$ 1500,00 em diante : aumento de 5% Após o aumento ser realizado, informe na tela:
    o salário antes do reajuste;
    o percentual de aumento aplicado;
    o valor do aumento;
    o novo salário, após o aumento.
    

In [ ]:
salario_antigo = float(input('Digite o salário do colaborador: R$'))

if salario_antigo <= 280:
  percentual_aumento = 0.2

elif 280 < salario_antigo <= 700:
  percentual_aumento = 0.15

elif 700 < salario_antigo <= 1500:
  percentual_aumento = 0.1

else:
  percentual_aumento = 0.05

valor_aumento = salario_antigo * percentual_aumento
novo_salario = valor_aumento + salario_antigo

print(f'''
o salário antes do reajuste = R${salario_antigo:.2f}
o percentual de aumento aplicado = {percentual_aumento:.0%}
o valor do aumento = R${valor_aumento:.2f}
o novo salário, após o aumento = {novo_salario:.2f}
 ''')

Exercício 12

Faça um programa para o cálculo de uma folha de pagamento, sabendo que os descontos são do Imposto de Renda, que depende do salário bruto (conforme tabela abaixo) e 3% para o Sindicato e que o FGTS corresponde a 11% do Salário Bruto, mas não é descontado (é a empresa que deposita). O Salário Líquido corresponde ao Salário Bruto menos os descontos. O programa deverá pedir ao usuário o valor da sua hora e a quantidade de horas trabalhadas no mês.

    Desconto do IR:
    Salário Bruto até 900 (inclusive) - isento
    Salário Bruto até 1500 (inclusive) - desconto de 5%
    Salário Bruto até 2500 (inclusive) - desconto de 10%
    Salário Bruto acima de 2500 - desconto de 20% Imprima na tela as informações, dispostas conforme o exemplo abaixo. No exemplo o valor da hora é 5 e a quantidade de hora é 220.

            Salário Bruto: (5 * 220)        : R$ 1100,00
            (-) IR (5%)                     : R$   55,00  
            (-) INSS ( 10%)                 : R$  110,00
            FGTS (11%)                      : R$  121,00
            Total de descontos              : R$  165,00
            Salário Liquido                 : R$  935,00

In [ ]:
sindicato_percent = 0.03
fgts_percent = 0.11
inss_percent = 0.10
valor_hora = float(input('Digite o valor hora: '))
quantidade_hora = int(input('Digite a quantidade hora: '))
salario_bruto = valor_hora * quantidade_hora

if salario_bruto <= 900:
  ir_percent = 0

elif 900 < salario_bruto <= 1500:
  ir_percent = 0.05

elif 1500 < salario_bruto <= 2500:
  ir_percent = 0.1
  
else:
  ir_percent = 0.2

ir_desconto = ir_percent * salario_bruto
sindicato_desconto = sindicato_percent * salario_bruto
inss_desconto = inss_percent * salario_bruto

fgts = fgts_percent * salario_bruto
inss_desconto = inss_percent * salario_bruto
total_descontos = ir_desconto + sindicato_desconto + inss_desconto
salario_liquido = salario_bruto - total_descontos
print(f'''
CÁLCULO SALÁRIO

 Salário Bruto: ({valor_hora:.0f} * {quantidade_hora:.0f}) : R${salario_bruto:.2f}
(-) IR ({ir_percent:.0%}){' '*15}: R${ir_desconto:.2f}
(-) INSS ({inss_percent:.0%}){' '*12}: R${inss_desconto:.2f}
(-) Sindicato ({sindicato_percent:.0%}){' '*8}: R${sindicato_desconto:.2f}
 FGTS ({fgts_percent:.0%}){' '*15}: R${fgts:.2f}
 Total de descontos{' '*7}: R${total_descontos:.2f}
 Salário Líquido{' '*10}: R${salario_liquido:.2f}
''')

Exercício 13

Faça um Programa que leia um número e exiba o dia correspondente da semana. (1-Domingo, 2- Segunda, etc.), se digitar outro valor deve aparecer valor inválido. 

In [ ]:
num = int(input('Digite o número da semana: '))

if num == 1:
  print('Domingo')

elif num == 2:
  print('Segunda')

elif num == 3:
  print('Terça')

elif num == 4:
  print('Quarta')

elif num == 5:
  print('Quinta')

elif num == 6:
  print('Sexta')

elif num == 7:
  print('Sábado')
  
else:
  print('Valor Inválido')

Exercício 14

Faça um programa que lê as duas notas parciais obtidas por um aluno numa disciplina ao longo de um semestre, e calcule a sua média. A atribuição de conceitos obedece à tabela abaixo:

      Média de Aproveitamento  Conceito
      Entre 9.0 e 10.0        A
      Entre 7.5 e 9.0         B
      Entre 6.0 e 7.5         C
      Entre 4.0 e 6.0         D
      Entre 4.0 e zero        E

    O algoritmo deve mostrar na tela as notas, a média, o conceito correspondente e a mensagem “APROVADO” se o conceito for A, B ou C ou “REPROVADO” se o conceito for D ou E. 

In [ ]:
nota_1 = float(input('Digite a primeira nota parcial: '))
nota_2 = float(input('Digite a segunda nota parcial: '))
media = (nota_1 + nota_2) / 2

if 9 < media <= 10:
  conceito = 'A'

elif 7.5 < media <= 9:
  conceito = 'B'

elif 6 < media <= 7.5:
  conceito = 'C'

elif 4 < media <= 6:
  conceito = 'D'
  
elif 0 <= media <= 4:
  conceito = 'E'

if conceito == 'A' or conceito=='B' or conceito == 'C':
  aprovacao = 'Aprovado'
else:
  aprovacao = 'Reprovado'

print(f'''
nota 01 = {nota_1:.2f}
nota 02 = {nota_2:.2f}
media = {media:.2f}
conceito = {conceito}  "{aprovacao}"
''')

Exercício 15

Faça um Programa que peça os 3 lados de um triângulo. O programa deverá informar se os valores podem ser um triângulo. Indique, caso os lados formem um triângulo, se o mesmo é: equilátero, isósceles ou escaleno.

    Dicas:
    Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
    Triângulo Equilátero: três lados iguais;
    Triângulo Isósceles: quaisquer dois lados iguais;
    Triângulo Escaleno: três lados diferentes; 
    

In [ ]:
lado_1 = float(input('Digite o primeiro lado do triângulo: '))
lado_2 = float(input('Digite o segundo lado do triângulo: '))
lado_3 = float(input('Digite o terceiro lado do triângulo: '))

if (lado_1 + lado_2  > lado_3) and (lado_1 + lado_3  > lado_2) and (lado_2 + lado_3  > lado_1):
  if lado_1 == lado_2  == lado_3:
    tipo_triangulo = 'Equilátero'
  elif lado_1 != lado_2  != lado_3:
    tipo_triangulo = 'Escaleno'
  else:
    tipo_triangulo = 'Isósceles'
  print(f'É um triângulo {tipo_triangulo}')
else:
  print('Não é um triângulo')

Exercício 16

Faça um programa que calcule as raízes de uma equação do segundo grau, na forma ax2 + bx + c. O programa deverá pedir os valores de a, b e c e fazer as consistências, informando ao usuário nas seguintes situações:

    Se o usuário informar o valor de A igual a zero, a equação não é do segundo grau e o programa não deve fazer pedir os demais valores, sendo encerrado;
    Se o delta calculado for negativo, a equação não possui raizes reais. Informe ao usuário e encerre o programa;
    Se o delta calculado for igual a zero a equação possui apenas uma raiz real; informe-a ao usuário;
    Se o delta for positivo, a equação possui duas raiz reais; informe-as ao usuário; 

In [ ]:
a = float(input('Digite o valor de a: '))
if a == 0:

  print('A equação não é do segundo grau')

else:
  b = float(input('Digite o valor de b: '))
  c = float(input('Digite o valor de c: '))
  delta = (b ** 2) - (4*a*c)

  if delta < 0:
    print('A equação não possui raizes reais.')

  elif delta == 0:
    print('A equação uma raiz real.')
    
  else:
    print('A equação possui duas raizes reais.')


Exercício 17

Faça um Programa que peça um número correspondente a um determinado ano e em seguida informe se este ano é ou não bissexto.

In [ ]:
ano = int(input('Digite o ano: '))
if ano % 4 == 0:
  print('Ano bissexto')
  
else:
  print('Ano  NÃO bissexto')

Exercício 18

Faça um Programa que peça uma data no formato dd/mm/aaaa e determine se a mesma é uma data válida. 

In [ ]:
is_data_valida = is_data_formato_valida = is_ano_valido = is_mes_valido = is_dia_valido  = False

data = input(f'Digite uma data dd/mm/aaaa: ')

try:
  dia = int(data[0:2])
  mes = int(data[3:5])
  ano = int(data[6:])

except ValueError:
  print(f'Data inválida! {data[0:2]} / {data[3:5]} / {data[6:]} ')

else:
  is_ano_valido = 1 <= ano <= 9999
  
  if is_ano_valido:
    is_ano_bissexto = ano % 4 == 0
    is_mes_valido = 1 <= mes <= 12
    
    if is_mes_valido:
      is_mes_31_dias = mes in (1, 3, 5, 7, 8, 10, 12)
      is_mes_30_dias = mes in (4, 6, 9, 11)
      is_mes_28_dias = mes == 2 and not is_ano_bissexto
      is_mes_29_dias = mes == 2 and is_ano_bissexto
      
      is_dia_valido = ((is_mes_28_dias and dia in range(1,29)) or 
                       (is_mes_29_dias and dia in range(1,30)) or 
                       (is_mes_30_dias and dia in range(1,31)) or 
                       (is_mes_31_dias and dia in range(1,32)))
      
  is_data_formato_valida = len(data) == 10 and data[2] == '/' and data[5] == '/'
  is_data_valida = is_data_formato_valida and is_ano_valido and is_mes_valido and is_dia_valido
  
  msg = 'Data Iválida! '
  if is_data_valida:
    msg = 'Data Válida: '
   
  print(f'{msg} {str(dia).zfill(2)}/{str(mes).zfill(2)}/{str(ano).zfill(4)}')

    

Exercício 19

Faça um Programa que leia um número inteiro menor que 1000 e imprima a quantidade de centenas, dezenas e unidades do mesmo.

    Observando os termos no plural a colocação do "e", da vírgula entre outros. Exemplo:
    326 = 3 centenas, 2 dezenas e 6 unidades
    12 = 1 dezena e 2 unidades Testar com: 326, 300, 100, 320, 310,305, 301, 101, 311, 111, 25, 20, 10, 21, 11, 1, 7 e 16 

In [ ]:
nums = [326, 300, 100, 320, 310, 305, 301, 101, 311, 111, 25, 20, 10, 21, 11, 1, 7 , 16, 276]

for num in nums:
  numero = num
  centena = num // 100
  dezena = num % 100
  num = num % 10
  unidade = num // 1
  centena_msg=  dezena_msg = unidade_msg = ''
  if centena > 0:
    centena_msg = f"{centena} {'centenas' if centena > 1 else 'centena'}"
  if dezena > 0:
    dezena_msg = f"{', ' if centena_msg!='' else ''} {dezena} {'dezenas' if dezena > 1 else 'dezena'}"
  if unidade > 0:
    unidade_msg = f"{' e'if (centena_msg!='' or dezena_msg!='') else ''} {unidade} {'unidades' if unidade > 1 else 'unidade'}"
  print(f'{numero} = {centena_msg}{dezena_msg}{unidade_msg}.')


Execício 21

Faça um Programa para um caixa eletrônico. O programa deverá perguntar ao usuário a valor do saque e depois informar quantas notas de cada valor serão fornecidas. As notas disponíveis serão as de 1, 5, 10, 50 e 100 reais. O valor mínimo é de 10 reais e o máximo de 600 reais. O programa não deve se preocupar com a quantidade de notas existentes na máquina.

    Exemplo 1: Para sacar a quantia de 256 reais, o programa fornece duas notas de 100, uma nota de 50, uma nota de 5 e uma nota de 1;
    Exemplo 2: Para sacar a quantia de 399 reais, o programa fornece três notas de 100, uma nota de 50, quatro notas de 10, uma nota de 5 e quatro notas de 1. 

In [ ]:
valor_saque = int(input('Digite o valor a ser sacado: '))

contagem = valor_saque
notas_cem = contagem // 100
contagem = contagem % 100
notas_cinquenta = contagem // 50
contagem = contagem % 50
notas_dez = contagem // 10
contagem = contagem % 10
notas_cinco = contagem // 5
notas_um = contagem % 5


if notas_cem <= 0:
  msg_100 = ''
else:
  msg_100 = f'{notas_cem} {"notas" if notas_cem >1 else "nota"} de R$100.00'

if notas_cinquenta <= 0:
  msg_50 = ''
else:
  msg_50 = f'{", " if msg_100!= ""else ""}{notas_cinquenta} {"notas" if notas_cinquenta >1 else "nota"} de R$50.00'

if notas_dez <= 0:
  msg_10 = ''
else:
  msg_10 = f'{", " if (msg_100!= "" or msg_50!= "")else ""}{notas_dez} {"notas" if notas_dez >1 else "nota"} de R$10.00'

if notas_cinco <= 0:
  msg_5 = ''
else:
  msg_5 = f'{", " if (msg_100!= "" or msg_50!= "" or msg_10!= "")else ""}{notas_cinco} {"notas" if notas_cinco >1 else "nota"} de R$5.00'

if notas_um <= 0:
  msg_1 = ''
else:
  msg_1 = f'{" e " if (msg_100!= "" or msg_50!= "" or msg_10!= "" or msg_5!= "")else ""}{notas_um} {"notas" if notas_um >1 else "nota"} de R$1.00'

print(f' Para sacar a quantia de {valor_saque} reais, o programa fornece: {msg_100}{msg_50}{msg_10}{msg_5}{msg_1}.')

In [ ]:
6 // 5

In [ ]:
help(divmod)

# [Estrutura de Repetição](https://wiki.python.org.br/EstruturaDeRepeticao)


Exercício 1

Faça um programa que peça uma nota, entre zero e dez. Mostre uma mensagem caso o valor seja inválido e continue pedindo até que o usuário informe um valor válido. 

In [ ]:

while True:
  nota = input('Digite uma nota entre zero e dez: ')
  try:
    nota_valida = float(nota)
  except ValueError:
    print('valor digitado inválido!')
  else:
    if 0 <= nota_valida <= 10:
      break
    else:
      print('Digite um número entre 0 e 10')

Digite uma nota entre zero e dez: 10


Exercício 2

Faça um programa que leia um nome de usuário e a sua senha e não aceite a senha igual ao nome do usuário, mostrando uma mensagem de erro e voltando a pedir as informações.

In [ ]:
while True:
  nome = input('Digite seu nome: ')
  senha = input('Digite sua senha: ')

  if nome == senha:
    print('Nome não pode ser igual a senha, voltano a pedir informações...')
  else:
    print('Nome e senha armazenados com sucesso!.')
    break

Digite seu nome: aaa
Digite sua senha: aaa
Nome não pode ser igual a senha, voltano a pedir informações...
Digite seu nome: bbb
Digite sua senha: bbb
Nome não pode ser igual a senha, voltano a pedir informações...
Digite seu nome: abc
Digite sua senha: bca
Nome e senha armazenados com sucesso!


Faça um programa que leia e valide as seguintes informações:

    Nome: maior que 3 caracteres;
    Idade: entre 0 e 150;
    Salário: maior que zero;
    Sexo: 'f' ou 'm';
    Estado Civil: 's', 'c', 'v', 'd'; 

In [ ]:
while True:
  nome = input('Digite o seu nome: ')
  if len(nome) >=3:
    break
  else:
    print('seu nome precisa ser maior que 3 caracteres...')
while True:
  idade = input('digite sua idade: ')
  try:
    idade = int(idade)
  except ValueError:
    print('Valor para idade inválido!')
  else:
    if 0 < idade < 150:
      break
    else:
      print('Idade precisa estar entre 0 e 150')
while True:
  salario = input('Digite seu salário: ')
  try:
    salario = float(salario)
  except ValueError:
    print('Salário inválido!')
  else:
    if salario > 0:
      break
    else:
      print('Salário precisa ser maior que zero!')
while True:
  sexo = (input('Digite seu sexo( f ou m)')).lower()
  if sexo == 'f' or sexo == 'm':
    break
  else:
    print('Sexo inválido!')
while True:
  estado_civil = input('Digite o Estado Civil ("s", "c", "v", "d") ').lower()
  if estado_civil in ['s', 'c', 'v', 'd']:
    break
  else:
    print('Estado Civil Inválido')

    

Digite o seu nome: zx
seu nome precisa ser maior que 3 caracteres...
Digite o seu nome: zxcz
digite sua idade: -3
Idade precisa estar entre 0 e 150
digite sua idade: 1610
Idade precisa estar entre 0 e 150
digite sua idade: 101
Digite seu salário: 1200
Digite seu sexo( f ou m)d
Sexo inválido!
Digite seu sexo( f ou m)d
Sexo inválido!
Digite seu sexo( f ou m)F
Digite o Estado Civil ("s", "c", "v", "d") Q
Estado Civil Inválido
Digite o Estado Civil ("s", "c", "v", "d") Y
Estado Civil Inválido
Digite o Estado Civil ("s", "c", "v", "d") S


Exercício 4

Supondo que a população de um país A seja da ordem de 80000 habitantes com uma taxa anual de crescimento de 3% e que a população de B seja 200000 habitantes com uma taxa de crescimento de 1.5%. Faça um programa que calcule e escreva o número de anos necessários para que a população do país A ultrapasse ou iguale a população do país B, mantidas as taxas de crescimento.


In [ ]:
habitantes_a = 80_000
habitantes_b = 200_000
cresc_tax_a = 1.03
cresc_tax_b = 1.015
anos = 0
while habitantes_a < habitantes_b:
#  print(f'''***** Crescimento Habitantes A e B
# Ano: {anos}
# Habitantes A: {habitantes_a}
# Habitantes B: {habitantes_b}
# ''')

  habitantes_a = int(habitantes_a * cresc_tax_a)
  habitantes_b = int(habitantes_b * cresc_tax_b)
  anos += 1

print(f'''***** Crescimento Habitantes A e B
Ano: {anos}
Habitantes A: {habitantes_a}
Habitantes B: {habitantes_b}
''')

***** Crescimento Habitantes A e B
Ano: 63
Habitantes A: 514954
Habitantes B: 510908



Exercício 5

Altere o programa anterior permitindo ao usuário informar as populações e as taxas de crescimento iniciais. Valide a entrada e permita repetir a operação. 

In [ ]:

while True: 
  try:
    habitantes_a = int(input('Digite o NÚMERO DE HABITANTES da do país que tem MENOR população: '))
  
  except ValueError:
    print('Valor Inválido!')
  
  else:
    break

while True:  
  try:
    habitantes_b = int(input('Digite o NÚMERO DE HABITANTES da do país que tem MAIOR população: '))
  
  except ValueError:
    print('Valor Inválido!')
  
  else:
    if habitantes_a < habitantes_b:
      break

    else:
      print(f'O NÚMERO DE HABITANTES precisa ser maior que {habitantes_a}!')

while True: 
  try:
    tax_cresc_a = float(input('Digite a TAXA DE CRESCIMENTO ( % ) da do país que tem MENOR população: '))
  
  except ValueError:
    print('Valor Inválido!')
  
  else:
    if 0 < tax_cresc_a <= 100:
      break
    
    else:
      print('A TAXA DE CRESCIMENTO precisa ser um valor ENTRE ZERO e CEM')


while True: 
  try:
    tax_cresc_b = float(input('Digite a TAXA DE CRESCIMENTO ( % ) da do país que tem MAIOR população: '))
  
  except ValueError:
    print('Valor Inválido!')
  
  else:
    if 0 < tax_cresc_b < tax_cresc_a:
      break
      
    else:
      print(f'A TAXA DE CRESCIMENTO precisa ser um valor MAIOR que ZERO e MENOR que {tax_cresc_a}')

anos = 0
tax_cresc_a /= 100
tax_cresc_b /= 100

while habitantes_a < habitantes_b:
#  print(f'''***** Crescimento Habitantes A e B
# Ano: {anos}
# Habitantes A: {habitantes_a}
# Habitantes B: {habitantes_b}
# ''')

  habitantes_a = int(habitantes_a + habitantes_a * tax_cresc_a)
  habitantes_b = int(habitantes_b + habitantes_b * tax_cresc_b)
  anos += 1

print(f'''
***** Crescimento Habitantes A e B
Ano: {anos}
Habitantes A: {habitantes_a}
Habitantes B: {habitantes_b}
''')

Digite o NÚMERO DE HABITANTES da do país que tem MENOR população: 80000
Digite o NÚMERO DE HABITANTES da do país que tem MAIOR população: 200000
Digite a TAXA DE CRESCIMENTO ( % ) da do país que tem MENOR população: 3
Digite a TAXA DE CRESCIMENTO ( % ) da do país que tem MAIOR população: 1.5

***** Crescimento Habitantes A e B
Ano: 63
Habitantes A: 514954
Habitantes B: 510914



Exercício 6

Faça um programa que imprima na tela os números de 1 a 20, um abaixo do outro. Depois modifique o programa para que ele mostre os números um ao lado do outro. 

In [ ]:

for x in range(1, 121):
  print(x)

msg=''
for x in range (1, 121):
  msg += str(x) 
  msg+=' '
print(msg)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 


Exercício 7

Faça um programa que leia 5 números e informe o maior número. 

In [ ]:
num = float(input('Digite um número: '))

for _ in range(4):
  try:
    maximo = max(num, float(input('Digite um número: ')))
  
  except ValueError:
    print('Valor Inválido!')

print(maximo)

Digite um número: 3
Digite um número: sdfsd
Valor Inválido!
Digite um número: 23
Digite um número: 32
Digite um número: 32
32.0


Exercício 8

Faça um programa que leia 5 números e informe a soma e a média dos números. 

In [ ]:
soma = media = num = float(input('Digite um número: '))
for n in range(2, 6):
  soma +=  float(input('Digite um número: '))
  media = soma / n
  print(f'''soma=  {soma} e média = {media}''')

      
  
      

Digite um número: 2
Digite um número: 3
soma=  5.0 e média = 2.5
Digite um número: 4
soma=  9.0 e média = 3.0
Digite um número: 5
soma=  14.0 e média = 3.5
Digite um número: 5
soma=  19.0 e média = 3.8


Exercício 9

Faça um programa que imprima na tela apenas os números ímpares entre 1 e 50. 

In [ ]:
for x in range(1, 50):
  if x % 2 != 0:
    print(x)
    

1
3
5
7
9
11
13
15
17
19
21
23
25
27
29
31
33
35
37
39
41
43
45
47
49


# [Listas](https://wiki.python.org.br/ExerciciosListas)

Exercício 1

Faça um Programa que leia um vetor de 5 números inteiros e mostre-os.

In [ ]:
lista = []

for _ in range(5):
  while True:
    try:
      num = int(input('Digite um número: '))
    
    except ValueError:
      print('Valor Inválido!')
    
    else:
      lista.append(num)
      break

for x in lista:
  print(x)

Digite um número: fgdffg
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: fsdfsdf
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: sfdfsdf
Valor Inválido!
Digite um número: 3
Digite um número: sdfsdf
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: 6.6
Digite um número: sdfsdf
Valor Inválido!
Digite um número: sdfsdf
Valor Inválido!
Digite um número: 7.9
Digite um número: sfdsdfsd
Valor Inválido!
Digite um número: 5.6
Digite um número: sdfsdf
Valor Inválido!
Digite um número: 3.6
3.0
6.6
7.9
5.6
3.6


Exercício 2

Faça um Programa que leia um vetor de 10 números reais e mostre-os na ordem inversa. 

In [ ]:
lista = []
for _ in range(10):
  while True:
    try:
      num = float(input('Digite um número real: '))
    
    except ValueError:
      print('Valor Inválido!')
   
    else:
      lista.append(num)
      break

print(lista[::-1])

Digite um número real: dfsdf
Valor Inválido!
Digite um número real: sdfsdf
Valor Inválido!
Digite um número real: 1
Digite um número real: 2
Digite um número real: 3
Digite um número real: 4.5
Digite um número real: 5.3
Digite um número real: 56.898
Digite um número real: 7
Digite um número real: .9
Digite um número real: -9.3
Digite um número real: -56.9
[-56.9, -9.3, 0.9, 7.0, 56.898, 5.3, 4.5, 3.0, 2.0, 1.0]


Exercício 3

Faça um Programa que leia 4 notas, mostre as notas e a média na tela. 

In [ ]:
notas =[]
for _ in range(4):
  while True:
    try:
      nota = float(input('Digite a nota: '))
    
    except:
      print('Valor Inválido!')
    
    else:
      notas.append(nota)
      break

soma = sum(notas)
media = soma / len(notas)

print(f'***Notas e Média')
for c, v in enumerate(notas):
  print(f'nota {c+1}= {v:.2f}')

print(f'média = {media:.2f}')

Digite a nota: 5.5
Digite a nota: 6.6
Digite a nota: 7.7
Digite a nota: 8.8
***Notas e Média
nota 1= 5.50
nota 2= 6.60
nota 3= 7.70
nota 4= 8.80
média = 7.15


Exercício 4

Faça um Programa que leia um vetor de 10 caracteres, e diga quantas consoantes foram lidas. Imprima as consoantes. 

In [ ]:
lista_consoantes = 'bcdfghjklmnpqrstwvxyz'
lista_consoantes = lista_consoantes + lista_consoantes.upper()
while True:
  vector_10 = input('Digite um 10 CARACTERES na tela: ')
  
  if len(vector_10) == 10:
    break
  
  else:
    print(f'Total caracteres inválidos! {len(vector_10)}')

consoantes = []
for x in  vector_10:
  if x in lista_consoantes:
    consoantes.append(x)

print(f'''
vetor caracteres: {vector_10}
total de consoantes: {len(consoantes)}
consoantes: {' '.join(consoantes)}''')




Digite um 10 CARACTERES na tela: jXlua63!op

vetor caracteres: jXlua63!op
total de consoantes: 4
consoantes: j X l p


Exercício 5

Faça um Programa que leia 20 números inteiros e armazene-os num vetor. Armazene os números pares no vetor PAR e os números IMPARES no vetor impar. Imprima os três vetores. 

In [ ]:
numeros = []
pares = []
impares = []
for _ in range(20):
  while True:
    try:
      num  = int(input('Digite um número inteiro: '))
    
    except ValueError:
      print('Valor Inválido!')
    
    else:
      numeros.append(num)
      if num % 2 == 0:
        pares.append(num)
      else:
        impares.append(num)
      break

print(f'''
números: {numeros}
pares: {pares}
impares: {impares}
''')

Digite um número inteiro: 5
Digite um número inteiro: 12
Digite um número inteiro: 75
Digite um número inteiro: bvf
Valor Inválido!
Digite um número inteiro: 452
Digite um número inteiro: 4.6
Valor Inválido!
Digite um número inteiro: 75
Digite um número inteiro: 35
Digite um número inteiro: 75
Digite um número inteiro: -9
Digite um número inteiro: 12
Digite um número inteiro: -8
Digite um número inteiro: 12
Digite um número inteiro: 3
Digite um número inteiro: 7
Digite um número inteiro: 38
Digite um número inteiro: 53
Digite um número inteiro: 42
Digite um número inteiro: 423
Digite um número inteiro: 75
Digite um número inteiro: 64
Digite um número inteiro: 12

números: [5, 12, 75, 452, 75, 35, 75, -9, 12, -8, 12, 3, 7, 38, 53, 42, 423, 75, 64, 12]
pares: [12, 452, 12, -8, 12, 38, 42, 64, 12]
impares: [5, 75, 75, 35, 75, -9, 3, 7, 53, 423, 75]



Exercício 6

Faça um Programa que peça as quatro notas de 10 alunos, calcule e armazene num vetor a média de cada aluno, imprima o número de alunos com média maior ou igual a 7.0. 

In [ ]:
alunos = []
for x in range(10):
  alunos.append([])


for c, v in enumerate(alunos):
  for n in range(1,5):
    while True:
      try:
        nota = float(input(f'Aluno {c+1} -- digite a nota {n}: '))

      except ValueError:
        print('Valor Inválido!')

      else:
        if  0 <= nota <= 10:
          alunos[c].append(nota)
          break
        else:
          print('Nota precisa ser entre Zero e DEZ!')
  soma = sum(alunos[c])
  media = soma / len(alunos[c])
  alunos[c].append(media)  
print()
print("***ALUNOS COM MÉIDA >= 7")

for c, v in enumerate(alunos):
  if alunos[c][-1] >= 7:
    print(f'''
    Aluno {c+1}, média: {alunos[c][-1]}''')
print(),1
print(alunos)

Aluno 1 -- digite a nota 1: 6.3
Aluno 1 -- digite a nota 2: 73.
Nota precisa ser entre Zero e DEZ!
Aluno 1 -- digite a nota 2: 7.6
Aluno 1 -- digite a nota 3: 3.6
Aluno 1 -- digite a nota 4: 2.3
Aluno 2 -- digite a nota 1: 7.6
Aluno 2 -- digite a nota 2: 2.5
Aluno 2 -- digite a nota 3: 10
Aluno 2 -- digite a nota 4: 10
Aluno 3 -- digite a nota 1: 9.8
Aluno 3 -- digite a nota 2: 7
Aluno 3 -- digite a nota 3: 8.4
Aluno 3 -- digite a nota 4: 2.3
Aluno 4 -- digite a nota 1: 9.5
Aluno 4 -- digite a nota 2: 7.9
Aluno 4 -- digite a nota 3: 9.6
Aluno 4 -- digite a nota 4: 5.8
Aluno 5 -- digite a nota 1: 9.6
Aluno 5 -- digite a nota 2: 3.6
Aluno 5 -- digite a nota 3: 8.9
Aluno 5 -- digite a nota 4: 9.6
Aluno 6 -- digite a nota 1: 7.8
Aluno 6 -- digite a nota 2: 8.9
Aluno 6 -- digite a nota 3: 5
Aluno 6 -- digite a nota 4: 7
Aluno 7 -- digite a nota 1: 7
Aluno 7 -- digite a nota 2: 8
Aluno 7 -- digite a nota 3: 9
Aluno 7 -- digite a nota 4: 1
Aluno 8 -- digite a nota 1: 3
Aluno 8 -- digite a not

Exercícios 7

Faça um Programa que leia um vetor de 5 números inteiros, mostre a soma, a multiplicação e os números. 

In [ ]:
import math
vetor = [1, 2, 3, 4, 5]

soma = 0
produto = 1

for x in vetor:
  soma += x
  produto *= x 

for x in vetor:
  print(x)

print(f'soma: {soma}')
print(f'produto: {produto}')

1
2
3
4
5
soma: 15
produto: 120


Exercício 8

Faça um Programa que peça a idade e a altura de 5 pessoas, armazene cada informação no seu respectivo vetor. Imprima a idade e a altura na ordem inversa a ordem lida. 

In [ ]:
pessoas = []
for k in range(5):
  while True:
    try:
      idade = int(input(f'Pessoa {k+1}: Digite a idade: '))
      altura = float(input(f'Pessoa {k+1}: Digite a altura: '))
    except ValueError:
      print('Valores Inválidos!')
    else:
      pessoas.append([idade, altura])
      break

print('**** Pessoas')
for k, v in enumerate(pessoas):
  print(f'Pessoa {k+1}: altura:{pessoas[k][-1]} idade:{pessoas[k][0]}')

Pessoa 1: Digite a idade: 10
Pessoa 1: Digite a altura: 1.23
Pessoa 2: Digite a idade: 15
Pessoa 2: Digite a altura: 1.68
Pessoa 3: Digite a idade: 45
Pessoa 3: Digite a altura: 2.1
Pessoa 4: Digite a idade: 39
Pessoa 4: Digite a altura: 1.98
Pessoa 5: Digite a idade: 103
Pessoa 5: Digite a altura: 1.82
**** Pessoas
Pessoa 1: altura:1.23 idade:10
Pessoa 2: altura:1.68 idade:15
Pessoa 3: altura:2.1 idade:45
Pessoa 4: altura:1.98 idade:39
Pessoa 5: altura:1.82 idade:103


Exercicio 9

Faça um Programa que leia um vetor A com 10 números inteiros, calcule e mostre a soma dos quadrados dos elementos do vetor. 

In [ ]:
import random
vetor_a = []
for _ in range(10):
  vetor_a.append(random.randint(0, 100))

print(vetor_a)

soma = 0
for x in vetor_a:
  quad = x**2
  soma += quad

print(f'soma dos quadrados dos elementos: {soma}')

[83, 62, 59, 81, 91, 63, 21, 10, 83, 26]
soma dos quadrados dos elementos: 41131


Exercício 10

Faça um Programa que leia dois vetores com 10 elementos cada. Gere um terceiro vetor de 20 elementos, cujos valores deverão ser compostos pelos elementos intercalados dos dois outros vetores.

In [ ]:
while True:
  elementos_a = input('Digite 10 elementos do PRIMEIRO vetor separados por vírgula: ')
  elementos_a = elementos_a.split(',')
  if len(elementos_a) == 10:
    break
  else:
    print('Quantidade de elementos inválida!')

while True:
  elementos_b = input('Digite 10 elementos do SEGUNDO vetor separados por vírgula: ')
  elementos_b = elementos_b.split(',')
  if len(elementos_b) == 10:
    break
  else:
    print('Quantidade de elementos inválida!')

elementos_intercalados = []
for x in range(10):
  elementos_intercalados.append(elementos_a[x])
  elementos_intercalados.a ppend(elementos_b[x])


print(f'''
***ELEMENTOS INTERCALADOS
a = {elementos_a}
b = {elementos_b}
elementos intercalados = {elementos_intercalados}
''')

Digite 10 elementos do PRIMEIRO vetor separados por vírgula: 1,3,5,7,9,11,13,15,17,19
Digite 10 elementos do SEGUNDO vetor separados por vírgula: 2,4,6,8,10,12,14,16,18,20

***ELEMENTOS INTERCALADOS
a = ['1', '3', '5', '7', '9', '11', '13', '15', '17', '19']
b = ['2', '4', '6', '8', '10', '12', '14', '16', '18', '20']
elementos intercalados = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20']



Exercício 11

Altere o programa anterior, intercalando 3 vetores de 10 elementos cada. 

In [ ]:
while True:
  elementos_a = input('Digite 10 elementos do PRIMEIRO vetor separados por vírgula: ')
  elementos_a = elementos_a.split(',')
  if len(elementos_a) == 10:
    break
  else:
    print('Quantidade de elementos inválida!')

while True:
  elementos_b = input('Digite 10 elementos do SEGUNDO vetor separados por vírgula: ')
  elementos_b = elementos_b.split(',')
  if len(elementos_b) == 10:
    break
  else:
    print('Quantidade de elementos inválida!')

while True:
  elementos_c = input('Digite 10 elementos do TERCEIRO vetor separados por vírgula: ')
  elementos_c = elementos_c.split(',')
  if len(elementos_c) == 10:
    break
  else:
    print('Quantidade de elementos inválida!')

elementos_intercalados = []
for x in range(10):
  elementos_intercalados.append(elementos_a[x])
  elementos_intercalados.append(elementos_b[x])
  elementos_intercalados.a,ppend(elementos_c[x])


print(f'''
***ELEMENTOS INTERCALADOS
a = {elementos_a}
b = {elementos_b}
b = {elementos_c}
elementos intercalados = {elementos_intercalados}
''')

Digite 10 elementos do PRIMEIRO vetor separados por vírgula: 1,2,3,4,5,6,7,8
Quantidade de elementos inválida!
Digite 10 elementos do PRIMEIRO vetor separados por vírgula: 1,2,3,4,5,6,7,8,9,10
Digite 10 elementos do SEGUNDO vetor separados por vírgula: a,b,c,d,e,f,g,h,i,j,k,
Quantidade de elementos inválida!
Digite 10 elementos do SEGUNDO vetor separados por vírgula: a,b,c,d,e,f,g,h,i,j
Digite 10 elementos do Terceiro vetor separados por vírgula: !,@,#,$,%,¨,&,*,(,)

***ELEMENTOS INTERCALADOS
a = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
b = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
b = ['!', '@', '#', '$', '%', '¨', '&', '*', '(', ')']
elementos intercalados = ['1', 'a', '!', '2', 'b', '@', '3', 'c', '#', '4', 'd', '$', '5', 'e', '%', '6', 'f', '¨', '7', 'g', '&', '8', 'h', '*', '9', 'i', '(', '10', 'j', ')']



Exercício 12

Foram anotadas as idades e alturas de 30 alunos. Faça um Programa que determine quantos alunos com mais de 13 anos possuem altura inferior à média de altura desses alunos. 

In [ ]:
import random
alunos = []
for _ in range(30):
  idade = random.randint(1, 27)
  
  if 0 <= idade <= 12:
    altura = random.uniform(1, 1.2)
  else:
    altura = random.uniform(1, 1.7)
  
  alunos.append([idade, altura])

print(alunos)

soma_alturas = 0

for k, v in enumerate(alunos):
  soma_alturas +=  v[1]

media = soma_alturas / len(alunos)

print('***ALUNOS MAIORES DE 13 ANOS COM ALTURA INFERIOR À MEDIA:')
print(f'média = {media}')
for k, v in enumerate(alunos):
  if  13 < v[0] and v[1]<= media:
    print(f'''
    aluno {k+1}: idade = {v[0]}; altura = {v[1]}
    ''')

[[4, 1.1619964157959983], [12, 1.105157236003559], [6, 1.1624939135650316], [7, 1.1137088213717006], [24, 1.5697913003834802], [26, 1.2383072857436643], [13, 1.2108350783133899], [5, 1.1945501601631618], [3, 1.1782406801371041], [6, 1.1697875151056623], [8, 1.0128563893646403], [19, 1.2221967934213536], [23, 1.2985738783440204], [15, 1.3891320073898767], [26, 1.494059846438546], [1, 1.0008398617146308], [4, 1.1604010229520436], [27, 1.4178986160114244], [17, 1.4757238023865007], [24, 1.1851978832857961], [16, 1.5046817392893979], [6, 1.1453345799496126], [9, 1.194049130222904], [26, 1.5390760913355157], [2, 1.117275354462044], [15, 1.4094842324023058], [5, 1.1959282888396263], [14, 1.676874252635876], [8, 1.1861077603934518], [4, 1.0731782408536203]]
***ALUNOS MAIORES DE 13 ANOS COM ALTURA INFERIOR À MEDIA:
média = 1.2601246059425315

    aluno 6: idade = 26; altura = 1.2383072857436643
    

    aluno 12: idade = 19; altura = 1.2221967934213536
    

    aluno 20: idade = 24; altura =

Exercícios 13

Faça um programa que receba a temperatura média de cada mês do ano e armazene-as em uma lista. Após isto, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram (mostrar o mês por extenso: 1 – Janeiro, 2 – Fevereiro, . . . ). 

In [ ]:
import random
meses = []
nome_meses = ['1 - Janeiro', '2 - Fevereiro', '3 - Março', '4 - Abril', '5 - Maio', '6 - Junho', '7 - Julho', '8 - Agosto', '9 - Setembro', '10 - Outubro', '11 - Novembro', '12 - Dezembro']
for _ in range(12):
  meses.append(random.uniform(24 , 47))
media = sum(meses) / len(meses)

print(meses)

print(f'''
***MESES ACIMA DA MÉDIA ANUAL
Média Anual: {media}

''')

for k, v in enumerate(meses):
  if v >=  media:
    print(f'{nome_meses[k]}, temperatura = {v:.2f}')



[26.569606862949833, 29.440528311028483, 46.50444753674866, 45.68232800372248, 25.968768911574973, 41.283476824602516, 46.3023333299451, 41.08830344524589, 32.612496327681384, 34.212413775863226, 25.761235456100625, 28.365384274848395]

***MESES ACIMA DA MÉDIA ANUAL
Média Anual: 35.315943588359296


3 - Março, temperatura = 46.50
4 - Abril, temperatura = 45.68
6 - Junho, temperatura = 41.28
7 - Julho, temperatura = 46.30
8 - Agosto, temperatura = 41.09


Exercício 14

Utilizando listas faça um programa que faça 5 perguntas para uma pessoa sobre um crime. As perguntas são:

    "Telefonou para a vítima?"
    "Esteve no local do crime?"
    "Mora perto da vítima?"
    "Devia para a vítima?"
    "Já trabalhou com a vítima?" 
    O programa deve no final emitir uma classificação sobre a participação da pessoa no crime. Se a pessoa responder positivamente a 2 questões ela deve ser classificada como "Suspeita", entre 3 e 4 como "Cúmplice" e 5 como "Assassino". Caso contrário, ele será classificado como "Inocente". 

In [ ]:
perguntas =["Telefonou para a vítima?", "Esteve no local do crime?", "Mora perto da vítima?", "Devia para a vítima?", "Já trabalhou com a vítima?" ]
respostas = []

for pergunta in perguntas:
  while True:
    resposta = input(f'{pergunta} ')
    
    if resposta.lower() in ['s', 'sim', 'y', 'yes']:
      respostas.append(1)
      break
    
    elif resposta.lower() in ['n', 'não', 'no']:
      respostas.append(0)
      break
    
    else:
      print('Resposta Inválida!')

num_positivas = sum(respostas)

print(respostas)

if num_positivas == 2:
  print('Suspeita')

elif 3 <= num_positivas <= 4:
  print('Cúmplice')

elif num_positivas == 5:
  print('Assassino')

else:
  print('Inocente')




Telefonou para a vítima? SIM
Esteve no local do crime? SIM
Mora perto da vítima? NÃO
Devia para a vítima? NÃO
Já trabalhou com a vítima? NÃO
[1, 1, 0, 0, 0]
Suspeita


Exercício 15

Faça um programa que leia um número indeterminado de valores, correspondentes a notas, encerrando a entrada de dados quando for informado um valor igual a -1 (que não deve ser armazenado). Após esta entrada de dados, faça:

    Mostre a quantidade de valores que foram lidos;
    Exiba todos os valores na ordem em que foram informados, um ao lado do outro;
    Exiba todos os valores na ordem inversa à que foram informados, um abaixo do outro;
    Calcule e mostre a soma dos valores;
    Calcule e mostre a média dos valores;
    Calcule e mostre a quantidade de valores acima da média calculada;
    Calcule e mostre a quantidade de valores abaixo de sete;
    Encerre o programa com uma mensagem; 

In [ ]:
print('*** Digite as notas, para finalizar digite o valor -1.')
notas = []
notas_abaixo_sete = []

while True:
  try:
    nota = float(input('Digite uma nota: '))
  
  except ValueError:
    print('Nota Inválida!')

  else:
    if nota == -1:
      break
    
    elif nota > 10 or nota < 0:
      print('Notas precisam ser de 0 até 10')
    
    else:
      notas.append(nota)

      if nota < 7:
        notas_abaixo_sete.append(nota)

tamanho = len(notas)
soma = sum(notas)
media = soma /tamanho
valores = [str(x) for x in notas]
print(f'''
quantidade de valores lidos = {tamanho}
valores na ordem informada = {" - ".join(valores)}
valores na ordem inversa informada:
''')

notas_acima_media = []

notas_inversas = notas[::-1]
for nota in notas_inversas:
  if media < nota:
    notas_acima_media.append(nota)
  print(nota)

print(f'''
soma dos valores = {soma}
média dos valaroes = {media}
notas acima da média = {notas_acima_media}
notas abaixo de 7 = {notas_abaixo_sete}

'Programa Finalizado!'
''')

*** Digite as notas, para finalizar digite o valor -1.
Digite uma nota: 10
Digite uma nota: 5.5
Digite uma nota: 7
Digite uma nota: 9
Digite uma nota: -1

quantidade de valores lidos = 4
valores na ordem informada = 10.0 - 5.5 - 7.0 - 9.0
valores na ordem inversa informada:

9.0
7.0
5.5
10.0

soma dos valores = 31.5
média dos valaroes = 7.875
notas acima da média = [9.0, 10.0]
notas abaixo de 7 = [5.5]

'Programa Finalizado!'



In [ ]:
list_x =[1,2,3,4,5,8,9]
list_z = list_x[::]
list_y = list_x
print(list_z, id(list_z))
print(list_y, id(list_y))
print(list_x, id(list_x))

[1, 2, 3, 4, 5, 8, 9] 140294841306080
[1, 2, 3, 4, 5, 8, 9] 140294842203904
[1, 2, 3, 4, 5, 8, 9] 140294842203904


Exercício 16

Utilize uma lista para resolver o problema a seguir. Uma empresa paga seus vendedores com base em comissões. O vendedor recebe \$200 por semana mais 9 por cento de suas vendas brutas daquela semana. Por exemplo, um vendedor que teve vendas brutas de \$3000 em uma semana recebe \$200 mais 9 por cento de \$3000, ou seja, um total de \$470. Escreva um programa (usando um array de contadores) que determine quantos vendedores receberam salários nos seguintes intervalos de valores:

    $200 - $299
    $300 - $399
    $400 - $499
    $500 - $599
    $600 - $699
    $700 - $799
    $800 - $899
    $900 - $999
    $1000 em diante 

Desafio: Crie ma fórmula para chegar na posição da lista a partir do salário, sem fazer vários ifs aninhados. 

In [ ]:
salarios = [200, 250, 320, 413, 516, 680, 791, 877, 999, 1000, 2000, 3000]
faixas = [ x // 100 - 2 if x < 1000 else 8 for x in salarios]
print(contagem_faixa_salarial)

indices = []
[indices.append(i) for i in faixas if i not in indices]
contagem = [faixas.count(i) for i in indices]
contagem

NameError: ignored

# [Exercícios Funções](https://wiki.python.org.br/ExerciciosFuncoes)

Exercício 1

Faça um programa para imprimir:

        1
        2   2
        3   3   3
        .....
        n   n   n   n   n   n  ... n

    para um n informado pelo usuário. Use uma função que receba um valor n inteiro e imprima até a n-ésima linha. 

In [ ]:
dir(str)


def line(num):
  return (f'{num} ' * num).strip()

def colum(num):
  ret_msg =''
  for x in range(1, num + 1):
    ret_msg += line(x) + '\n'
  return ret_msg

print(colum(10))
 


1
2 2
3 3 3
4 4 4 4
5 5 5 5 5
6 6 6 6 6 6
7 7 7 7 7 7 7
8 8 8 8 8 8 8 8
9 9 9 9 9 9 9 9 9
10 10 10 10 10 10 10 10 10 10



Exercício 2

Faça um programa para imprimir:

        1
        1   2
        1   2   3
        .....
        1   2   3   ...  n

    para um n informado pelo usuário. Use uma função que receba um valor n inteiro imprima até a n-ésima linha. 

In [ ]:
def coluna(num):
  msg = ''
  for x in range(1, num + 1):
    msg += str(x) + ' '
  return msg.strip()

def linha(num):
  msg = ''
  for x in range(1, num + 1):
    msg += coluna(x) + '\n'
  return msg


print(linha(10))

1
1 2
1 2 3
1 2 3 4
1 2 3 4 5
1 2 3 4 5 6
1 2 3 4 5 6 7
1 2 3 4 5 6 7 8
1 2 3 4 5 6 7 8 9
1 2 3 4 5 6 7 8 9 10



In [ ]:
help('rs'.rstrip)

Help on built-in function rstrip:

rstrip(chars=None, /) method of builtins.str instance
    Return a copy of the string with trailing whitespace removed.
    
    If chars is given and not None, remove characters in chars instead.



# [Exercícios com Strings](https://wiki.python.org.br/ExerciciosComStrings)

Exercício 1

Tamanho de strings. Faça um programa que leia 2 strings e informe o conteúdo delas seguido do seu comprimento. Informe também se as duas strings possuem o mesmo comprimento e são iguais ou diferentes no conteúdo.

    Compara duas strings
    String 1: Brasil Hexa 2006
    String 2: Brasil! Hexa 2006!
    Tamanho de "Brasil Hexa 2006": 16 caracteres
    Tamanho de "Brasil! Hexa 2006!": 18 caracteres
    As duas strings são de tamanhos diferentes.
    As duas strings possuem conteúdo diferente.

In [ ]:
def compara_string(string_1, string_2):
  tamanho_1 = len(string_1)
  tamanho_2 = len(string_2)
  is_tamanho_igual = tamanho_1 == tamanho_2
  is_conteudo_igual = string_1 == string_2
  
  print(f'''
  Compara duas strings
  String 1: {string_1}
  String 2: {string_2}
  Tamanho de "{string_1}": {tamanho_1} caracteres
  Tamanho de "{string_2}": {tamanho_2} caracteres
  As duas strings são de tamanhos {'iguais.' if is_tamanho_igual else 'diferentes.'}
  As duas strings possuem conteúdos {'iguais.' if is_conteudo_igual else 'diferentes.'}
  ''')





In [ ]:
compara_string('Brasil Hexa 2006','Brasil! Hexa 2006!')


  Compara duas strings
  String 1: Brasil Hexa 2006
  String 2: Brasil! Hexa 2006!
  Tamanho de "Brasil Hexa 2006": 16 caracteres
  Tamanho de "Brasil! Hexa 2006!": 18 caracteres
  As duas strings são de tamanhos diferentes.
  As duas strings possuem conteúdos diferentes.
  


Exercicio 2

Nome ao contrário em maiúsculas. Faça um programa que permita ao usuário digitar o seu nome e em seguida mostre o nome do usuário de trás para frente utilizando somente letras maiúsculas. Dica: lembre−se que ao informar o nome o usuário pode digitar letras maiúsculas ou minúsculas. 

In [ ]:
nome = 'bruno torres mendes da silva'
letras_reversas = ''.join(reversed(nome.upper()))
palavras_reversas = ' '.join([i for i in (nome.upper().split())[::-1]])
letras_reversas_palavras_fixas =' '.join([i[::-1] for i in (nome.upper().split())])
print(f'''
reverso por letras: {letras_reversas}
reverso por palavras: {palavras_reversas}
reverso por letras nas palavras: {letras_reversas_palavras_fixas}''')


reverso por letras: AVLIS AD SEDNEM SERROT ONURB
reverso por palavras: SILVA DA MENDES TORRES BRUNO
reverso por letras nas palavras: ONURB SERROT SEDNEM AD AVLIS


Exercício 3

Nome na vertical. Faça um programa que solicite o nome do usuário e imprima-o na vertical.

    F
    U
    L
    A
    N
    O

In [ ]:

nome = input('Digite seu nome: ')

for letra in nome:
  print(letra)

Digite seu nome: bruno torres
b
r
u
n
o
 
t
o
r
r
e
s


Exercício 4

Nome na vertical em escada. Modifique o programa anterior de forma a mostrar o nome em formato de escada.

    F
    FU
    FUL
    FULA
    FULAN
    FULANO

In [ ]:
nome = input('Digite seu nome: ')

for i in range(len(nome) + 1):
  print(nome[:i])

Digite seu nome: FULANO

F
FU
FUL
FULA
FULAN
FULANO


Exercício 

Nome na vertical em escada invertida. Altere o programa anterior de modo que a escada seja invertida.

    FULANO
    FULAN
    FULA
    FUL
    FU
    F

In [ ]:
nome = 'FULANO'

for i in range(len(nome)):
  print(nome[:(-i if i !=0 else None)])


FULANO
FULAN
FULA
FUL
FU
F


Exercício 6

Data por extenso. Faça um programa que solicite a data de nascimento (dd/mm/aaaa) do usuário e imprima a data com o nome do mês por extenso.

    Data de Nascimento: 29/10/1973
    Você nasceu em  29 de Outubro de 1973.

In [ ]:
def check_is_data_valida(data):
  is_ano_valido = is_mes_valido = is_dia_valido = False

  if len(data) == 10 and data[2] == '/' and data[5] == '/':
    try:
      dia = int(data_nascimento[:2])
      mes = int(data_nascimento[3:5])
      ano = int(data_nascimento[-4:])
    
    except ValueError:
      print('Valores Inválidos!')
    
    else:
      
      is_ano_valido = 0 <= ano <= 3000
      is_ano_bissexto = ano % 4 == 0
      is_mes_valido = 1 <= mes <= 12
      
      if is_mes_valido:
        finais_meses = [30, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        
        if is_ano_bissexto:
          finais_meses[1] = 29
        
        is_dia_valido = 1 <= dia <= finais_meses[mes - 1]
    
    finally:
      return is_ano_valido and is_mes_valido and is_dia_valido, (dia, mes, ano)
  
  else:
    return is_ano_valido and is_mes_valido and is_dia_valido, (0, 0, 0)

def get_data_extenso(data):
  meses = ['Janeiro', 'Fevereiro', 'Março', 'Abril', 'Maio', 'Junho', 'Julho', 'Agosto', 'Setembro', 'Outubro', 'Novembro', 'Dezembro']
  return f'{data[0]} de {meses[data[1] -1]} de {data[2]}'

while True:
  data_nascimento = input('Digite sua data de nascimento no formato dd/mm/aaaa: ')
  is_data_valida, data = check_is_data_valida(data_nascimento)
  
  if is_data_valida:
    print(f'Data de Nascimento: {data_nascimento}')
    print(f'Você nasceu em {get_data_extenso(data)}')
    break
  
  else:
    print(f'Data Inválida {data}')






Digite sua data de nascimento no formato dd/mm/aaaa: 12/12/1912
Data de Nascimento: 12/12/1912
Você nasceu em 12 de Dezembro de 1912


Exercicio 7

Conta espaços e vogais. Dado uma string com uma frase informada pelo usuário (incluindo espaços em branco), conte:

    quantos espaços em branco existem na frase.
    quantas vezes aparecem as vogais a, e, i, o, u. 

In [ ]:
frase = (input('Digite uma frase: ')).lower()
print(frase)
qtd_branco = frase.count(' ')
qtd_a = frase.count('a')
qtd_e = frase.count('e')
qtd_i = frase.count('i')
qtd_o = frase.count('o')
qtd_u = frase.count('u')

print(f'''
***Contador de Vogais e Espaços Vazios:

quantidade de espaços vazios: {qtd_branco}
quantidade de a: {qtd_a}
quantidade de e: {qtd_e}
quantidade de i: {qtd_i}
quantidade de o: {qtd_o}
quantidade de u: {qtd_u}
''')

Digite uma frase: Testando o programa contador de vogais
testando o programa contador de vogais

***Contador de Vogais e Espaços Vazios:

quantidade de espaços vazios: 5
quantidade de a: 5
quantidade de e: 2
quantidade de i: 1
quantidade de o: 6
quantidade de u: 0



Exercício 8

Palíndromo. Um palíndromo é uma seqüência de caracteres cuja leitura é idêntica se feita da direita para esquerda ou vice−versa. Por exemplo: OSSO e OVO são palíndromos. Em textos mais complexos os espaços e pontuação são ignorados. A frase SUBI NO ONIBUS é o exemplo de uma frase palíndroma onde os espaços foram ignorados. Faça um programa que leia uma seqüência de caracteres, mostre−a e diga se é um palíndromo ou não. 

In [ ]:
sequencia = ''.join((input('Digite uma sequência ')).lower().strip().split())
if sequencia == sequencia[::-1]:
  print('palíndromo')

else:
  print('não é palíndromo')

Digite uma sequência SUBI NO ONIBUS 
palíndromo


Exercício 9

Verificação de CPF. Desenvolva um programa que solicite a digitação de um número de CPF no formato xxx.xxx.xxx-xx e indique se é um número válido ou inválido através da validação dos dígitos verificadores e dos caracteres de formatação. 

In [ ]:
cpf_orig = input('Digite seu CPF: ')
is_formatacao_valida = len(cpf_orig) == 14 and cpf_orig[3] == '.' and cpf_orig[7] == '.' and cpf_orig[11] == '-'

if is_formatacao_valida:
  cpf = cpf_orig.replace('.', '')
  cpf = cpf.replace('-', '')
  
  try:
    cpf = int(cpf)
  
  except ValueError:
    print('Numeração Inválida!')
  
  else:
    print(f'''Cpf Válido! {cpf_orig}''')
else:
  print('Formatação Inválida!')


Digite seu CPF: 010.704.375-06
Cpf Válido! 010.704.375-06


Exercício 9

Número por extenso. Escreva um programa que solicite ao usuário a digitação de um número até 99 e imprima-o na tela por extenso. 

In [ ]:
restos = ['um', 'dois', 'três', 'quatro', 'cinco', 'seis', 'sete', 'oito', 'nove']
fora_padrao =['onze','doze', 'treze', 'quatorze', 'quinze', 'dezesseis', 'dezessete', 'dezoito', 'dezenove,']
dividendos = ['dez', 'vinte', 'trinta', 'quarenta', 'cinquenta', 'sessenta', 'setenta', 'oitenta', 'noventa']

while True:
  try:
    num = int(input('Digite um número inteiro de 0 até 99: '))
    
    if num == -1:
      print('fim')
      break
  
  except ValueError:
    print('Digitação Inválida!')
  
  else:
    if 0<= num <= 99:
      dividendo = num // 10
      resto = num % 10
      num_txt = ''

      if dividendo == 0 and resto == 0:
        num_txt = f'zero'

      elif dividendo == 0 and resto >= 1:
        num_txt =f'{restos[resto-1]}'

      elif dividendo == 1 and resto >= 1:
        num_txt =f'{fora_padrao[resto-1]}'
      
      elif dividendo >= 1 and resto == 0:
        num_txt =f'{dividendos[dividendo - 1]}'
      
      elif dividendo >= 1 and resto > 0:
        num_txt =f'{dividendos[dividendo - 1]} e {restos[resto-1]}'
      
      else:
        num_txt ='número não identificado!'
      
      print(num_txt)

    else:
      print('Número Inválido')



Digite um número inteiro de 0 até 99: 65
sessenta e cinco
Digite um número inteiro de 0 até 99: 46
quarenta e seis
Digite um número inteiro de 0 até 99: 15
quinze
Digite um número inteiro de 0 até 99: 33
trinta e três
Digite um número inteiro de 0 até 99: 55
cinquenta e cinco
Digite um número inteiro de 0 até 99: -1
fim


Jogo de Forca. Desenvolva um jogo da forca. O programa terá uma lista de palavras lidas de um arquivo texto e escolherá uma aleatoriamente. O jogador poderá errar 6 vezes antes de ser enforcado.

    Digite uma letra: A
    -> Você errou pela 1ª vez. Tente de novo!

    Digite uma letra: O
    A palavra é: _ _ _ _ O

    Digite uma letra: E
    A palavra é: _ E _ _ O

    Digite uma letra: S
    -> Você errou pela 2ª vez. Tente de novo!

In [ ]:
palavra_secreta = 'churrasco'
lista_secreta = list(palavra_secreta)
forca = [' _ '] * len(palavra_secreta)
letras_escolhidas = []
erro = 0
max_erros = 6

while True:
  letra_escolhida = input('Digite uma letra: ')
  
  if len(letra_escolhida) != 1:
    print('Digite apenas uma letra por vez!')

  elif letra_escolhida in letras_escolhidas:
    print('Esta letra já foi escolhida')
  
  else:
    letras_escolhidas.append(letra_escolhida)
    
    if letra_escolhida not in palavra_secreta:
      erro += 1
      
      if erro == max_erros:
        print(f'-> Você errou pela {erro}ª vez. GAME OVER! ')
        break
      
      else:
        print(f'-> Você errou pela {erro}ª vez. Tente de novo! ')
      
      
    else:
      for k, v in enumerate(lista_secreta):
        if letra_escolhida == v:
          forca[k] = lista_secreta[k]

      acertos = "".join(forca)

      print(f'A palavra é: {acertos}')

      if acertos ==  palavra_secreta:
        print('Você GANHOU!')
        break




Digite uma letra: rr
Digite apenas uma letra por vez!
Digite uma letra: c
A palavra é: c _  _  _  _  _  _ c _ 
Digite uma letra: h
A palavra é: ch _  _  _  _  _ c _ 
Digite uma letra: u
A palavra é: chu _  _  _  _ c _ 
Digite uma letra: r
A palavra é: churr _  _ c _ 
Digite uma letra: a
A palavra é: churra _ c _ 
Digite uma letra: s
A palavra é: churrasc _ 
Digite uma letra: a
Esta letra já foi escolhida
Digite uma letra: t
-> Você errou pela 1ª vez. Tente de novo! 
Digite uma letra: a
Esta letra já foi escolhida
Digite uma letra: e
-> Você errou pela 2ª vez. Tente de novo! 
Digite uma letra: o
A palavra é: churrasco
Você GANHOU!


In [ ]:
'rr' in 'churrasco'

True

In [ ]:
my = '123.123.123-123'

help(my.isdigit)


Help on built-in function isdigit:

isdigit() method of builtins.str instance
    Return True if the string is a digit string, False otherwise.
    
    A string is a digit string if all characters in the string are digits and there
    is at least one character in the string.



# [Exercícios de Arquivos](https://wiki.python.org.br/ExerciciosArquivos)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

### Exercício 1

Faça um programa que leia um arquivo texto contendo uma lista de endereços IP e gere um outro arquivo, contendo um relatório dos endereços IP válidos e inválidos.

    O arquivo de entrada possui o seguinte formato: 

200.135.80.9

192.168.1.1

8.35.67.74

257.32.4.5

85.345.1.2

1.2.3.4

9.8.234.5

192.168.0.256

    O arquivo de saída possui o seguinte formato: 

[Endereços válidos:]

200.135.80.9

192.168.1.1

8.35.67.74

1.2.3.4


[Endereços inválidos:]

257.32.4.5

85.345.1.2

9.8.234.5

192.168.0.256

In [ ]:
def validar_ip(ip: str):
  ip_partes = ip.split('.')
  
  if len(ip_partes) != 4 :
    return False

  try:
    for x in ip_partes:
      if not 0 <= int(x) <= 255:
        return False
      
    return True
    
  except ValueError:
    return False


In [ ]:

ips_validos = []
ips_invalidos = []

with open('sample_data/ips.txt','r') as arquivo:
  for linha in arquivo:
    ip = linha.strip()
    if validar_ip(ip):
      ips_validos.append(ip)
      print(f'{ip} válido')
    else:
      ips_invalidos.append(ip)
      print(f'{ip} inválido')


200.135.80.9 válido
192.168.1.1 válido
8.35.67.74 válido
257.32.4.5 inválido
85.345.1.2 inválido
1.2.3.4 válido
9.8.234.5 válido
192.168.0.256 inválido
192.168. inválido
255.255.255.255 válido
A.V.0.256 inválido
192.B. inválido


In [ ]:
with open('sample_data/ips_saida.txt','w') as arquivo:
  arquivo.writelines('[Endereços Válidos]\n')
  for ip in ips_validos:
    arquivo.writelines(f'{ip}\n')
  arquivo.writelines('\n')
  arquivo.writelines('\n')
  arquivo.writelines('[Endereços Inválidos]\n')
  for ip in ips_invalidos:
    arquivo.writelines(f'{ip}\n')

## Exercício 2

A ACME Inc., uma empresa de 500 funcionários, está tendo problemas de espaço em disco no seu servidor de arquivos. Para tentar resolver este problema, o Administrador de Rede precisa saber qual o espaço ocupado pelos usuários, e identificar os usuários com maior espaço ocupado. Através de um programa, baixado da Internet, ele conseguiu gerar o seguinte arquivo, chamado "usuarios.txt":

alexandre       456123789

anderson        1245698456

antonio         123456456

carlos          91257581

cesar           987458

rosemary        789456125

Neste arquivo, o nome do usuário possui 15 caracteres. A partir deste arquivo, você deve criar um programa que gere um relatório, chamado "relatório.txt", no seguinte formato:

ACME Inc.               Uso do espaço em disco pelos usuários
------------------------------------------------------------------------
Nr.  Usuário        Espaço utilizado     % do uso

1    alexandre       434,99 MB             16,85%

2    anderson       1187,99 MB             46,02%

3    antonio         117,73 MB              4,56%

4    carlos           87,03 MB              3,37%

5    cesar             0,94 MB              0,04%

6    rosemary        752,88 MB             29,16%

Espaço total ocupado: 2581,57 MB

Espaço médio ocupado: 430,26 MB

O arquivo de entrada deve ser lido uma única vez, e os dados armazenados em memória, caso sejam necessários, de forma a agilizar a execução do programa. A conversão da espaço ocupado em disco, de bytes para megabytes deverá ser feita através de uma função separada, que será chamada pelo programa principal. O cálculo do percentual de uso também deverá ser feito através de uma função, que será chamada pelo programa principal. 

In [ ]:
def converter_byte_to_megabyte(num):
  return int(num) / 1_048_576

In [ ]:
usuarios = []

with open('sample_data/usuarios.txt', 'r') as arquivo:
  for linha in arquivo:
    usuario = linha.strip()
    usuarios.append((usuario[:16],converter_byte_to_megabyte(usuario[16:])))

total_megas = sum([v for k, v in usuarios])
media_megas = total_megas / len(usuarios)
usuarios = [(k, v, v/total_megas) for (k, v) in usuarios]

print(usuarios)


[('alexandre       ', 434.9935426712036, 0.1684991436018679), ('anderson        ', 1187.990623474121, 0.4601801705680585), ('antonio         ', 117.73725128173828, 0.04560671381277526), ('carlos          ', 87.03001117706299, 0.033711954115329186), ('cesar           ', 0.9417133331298828, 0.00036478217395237254), ('rosemary        ', 752.884030342102, 0.29163723572801675)]


In [ ]:
just_c1 = 10
just_c2 = 15
just_c3 = 10
just_c4 = 20

titulo = f'''{'ACME Inc.'.ljust(20)}{'Uso do espaço em disco pelos usuários'}
{'-'*75}
'''

cabecalho_txt = f'''{'Nr.'.ljust(just_c1)}{'Usuário'.ljust(just_c2)}{'Espaço utilizado'.center(just_c3)}{'% do uso'.center(just_c4)}
'''

usuarios_format = []
for k, v in enumerate(usuarios):
  usuario_numero = f'{k}'.ljust(just_c1)
  usuario_nome = f'{v[0]}'.ljust(just_c2)
  usuario_espaco = f'{v[1]:.2f} MB'.replace('.',',').rjust(just_c3)
  usuario_espaco_percent = f'{v[2]:.2%}'.replace('.',',').rjust(just_c4)
  
  usuarios_format.append(f'''
{usuario_numero}{usuario_nome}{usuario_espaco}{usuario_espaco_percent}''') 

rodape = f'''


Espaço total ocupado: {total_megas:.2f} MB
Espaço médio ocupado: {media_megas:.2f} MB'''

saida_txt = f'''{titulo}{cabecalho_txt}{''.join(usuarios_format)}{rodape}'''
  
with open('sample_data/relatorio.txt','w') as arquivo:
  arquivo.writelines(saida_txt)
  


In [ ]:
fill ='_'
for n in range(1,11):
  print(f'The number is {n:0>5}')


The number is 00001
The number is 00002
The number is 00003
The number is 00004
The number is 00005
The number is 00006
The number is 00007
The number is 00008
The number is 00009
The number is 00010


In [ ]:
x = f'{"final":x<10} {25566.588585:<2.2%}oi'
x

'finalxxxxx 2556658.86%oi'

In [ ]:
help(divmod)

Help on built-in function divmod in module builtins:

divmod(x, y, /)
    Return the tuple (x//y, x%y).  Invariant: div*y + mod == x.



## [Exercício com Classe](https://wiki.python.org.br/ExerciciosClasses)

Exercício 1

Classe Bola: Crie uma classe que modele uma bola:

    Atributos: Cor, circunferência, material
    Métodos: trocaCor e mostraCor 

In [ ]:
class Bola:
  def __init__(self):
    self.cor = 'Azul'
    self.circunferência = 4
    self.material = 'papel'
  
  def troca_cor(self, cor):
    self.cor = cor

  def mostra_cor(self):
    return self.cor


Exercício 2

Classe Quadrado: Crie uma classe que modele um quadrado:

    Atributos: Tamanho do lado
    Métodos: Mudar valor do Lado, Retornar valor do Lado e calcular Área; 

In [ ]:
class Quadrado:
  def __init__(self):
    self.lado = 1
  
  def calcular_area(self):
    return self.lado**2
 

Exercício 3

Classe Retangulo: Crie uma classe que modele um retangulo:

    Atributos: LadoA, LadoB (ou Comprimento e Largura, ou Base e Altura, a escolher)
    Métodos: Mudar valor dos lados, Retornar valor dos lados, calcular Área e calcular Perímetro;
    Crie um programa que utilize esta classe. Ele deve pedir ao usuário que informe as medidades de um local. Depois, deve criar um objeto com as medidas e calcular a quantidade de pisos e de rodapés necessárias para o local. 

In [ ]:
class Retangulo:
  def __init__(self):
    self.lado_1 = 1
    self.lado_2 = 1
  
  def mudar_lado_1(self, valor):
    self.lado_1 = valor
  
  def mudar_lado_2(self, valor):
    self.lado_2 = valor
  
  def get_lado_1(self):
    return self.lado_1

  def get_lado_2(self):
    return self.lado_2
  
  def calcular_area(self):
    return self.lado_1 * self.lado_2

  def calcular_perimetro(self):
    return self.lado_1 * self.lado_2
    

Exercício 4

Classe Pessoa: Crie uma classe que modele uma pessoa:

    Atributos: nome, idade, peso e altura
    Métodos: Envelhercer, engordar, emagrecer, crescer. Obs: Por padrão, a cada ano que nossa pessoa envelhece, sendo a idade dela menor que 21 anos, ela deve crescer 0,5 cm. 

In [ ]:
class Pessoa:
  def __spec__
  def __init__(self):
    self.nome = None
    self.idade = 0
    self.peso = 0
    self.altura = 0
  
  def envelhecer(self, valor):
    if idade < 21:
      self.crescer(0.5)
    self.idade += valor
  
  def engordar(self, valor):
    self.peso += valor

  def emagrecer(self, valor):
    self.peso -= valor
  
  def crescer(self, valor):
    self.altura += valor

In [ ]:
p = Pessoa()
p.engordar(5)
print(p)

Exercicio 5

Classe Conta Corrente: Crie uma classe para implementar uma conta corrente. A classe deve possuir os seguintes atributos: número da conta, nome do correntista e saldo. Os métodos são os seguintes: alterarNome, depósito e saque; No construtor, saldo é opcional, com valor default zero e os demais atributos são obrigatórios. 

In [ ]:
class ContaCorrente:
  def __init__(self, nome, numero_conta, saldo = 0):
    self.nome = nome
    self.numero_conta = numero_conta
    self.saldo = 0
  
  def __str__(self):
    return f'''{self.nome},{self.numero_conta},{self.saldo}'''
  

if __name__ == '__main__':
  conta_1 = ContaCorrente('Bruno', '2.654.321-56')
  print(conta_1)

Bruno,2.654.321-56,0


Exercício 6

Classe TV: Faça um programa que simule um televisor criando-o como um objeto. O usuário deve ser capaz de informar o número do canal e aumentar ou diminuir o volume. Certifique-se de que o número do canal e o nível do volume permanecem dentro de faixas válidas. 

In [ ]:
class Televisao:
  def __init__(self):
    self._volume = 0
    self._canal = 3

  @property
  def canal(self):
    return self._canal

  @canal.setter
  def canal(self, value):
    if 2 <= value <= 11:
      self._canal = value
      print(f'sintonizando canal: {value}')
    
  
  @property
  def volume(self):
    return self._volume

  @volume.setter
  def volume(self, value):
    if 0 <= value <= 100:
      self._volume = value
      print(f'Ajustando Volume em: {value}')
    else:
      if value > 100:
        print('volume máximo')
      else:
        print('volume mínimo')

if __name__ == '__main__':
  tv = Televisao()
  
  print(tv.volume)

  for x in range(10):
    print(x)
    tv.volume = x
    


0
0
Ajustando Volume em: 0
1
Ajustando Volume em: 1
2
Ajustando Volume em: 2
3
Ajustando Volume em: 3
4
Ajustando Volume em: 4
5
Ajustando Volume em: 5
6
Ajustando Volume em: 6
7
Ajustando Volume em: 7
8
Ajustando Volume em: 8
9
Ajustando Volume em: 9


Exercício 10

Classe Bomba de Combustível: Faça um programa completo utilizando classes e métodos que:

    Possua uma classe chamada bombaCombustível, com no mínimo esses atributos:
        tipoCombustivel.
        valorLitro
        quantidadeCombustivel 
    Possua no mínimo esses métodos:
        abastecerPorValor( ) – método onde é informado o valor a ser abastecido e mostra a quantidade de litros que foi colocada no veículo
        abastecerPorLitro( ) – método onde é informado a quantidade em litros de combustível e mostra o valor a ser pago pelo cliente.
        alterarValor( ) – altera o valor do litro do combustível.
        alterarCombustivel( ) – altera o tipo do combustível.
        alterarQuantidadeCombustivel( ) – altera a quantidade de combustível restante na bomba. 
    OBS: Sempre que acontecer um abastecimento é necessário atualizar a quantidade de combustível total na bomba. 

In [ ]:
class BombaCombustivel:
    def __init__(self, tipo_combustivel: str, valor_litro: float, quantidade_combustivel: float):
        self._tipo_combustivel = None
        self._valor_litro = None
        self._quantidade_combustivel = None
        self.tipo_combustivel = tipo_combustivel
        self.valor_litro = valor_litro
        self.quantidade_combustivel = quantidade_combustivel

    @property
    def tipo_combustivel(self):
        return self._tipo_combustivel

    @tipo_combustivel.setter
    def tipo_combustivel(self, valor):
        self._tipo_combustivel = valor

    @property
    def valor_litro(self):
        return self._valor_litro

    @valor_litro.setter
    def valor_litro(self, valor):
        if not valor > 0:
            raise ValueError(f'Valor Litro Inválido! {valor}')
        self._valor_litro = valor

    @property
    def quantidade_combustivel(self):
        return self._quantidade_combustivel

    @quantidade_combustivel.setter
    def quantidade_combustivel(self, valor):
        if valor <= 0:
            self._quantidade_combustivel = 0
        else:
            self._quantidade_combustivel = valor

    def abastecer_por_valor(self, valor):
        print(f'\nAbastecendo {self.tipo_combustivel} por valor: R${valor:.2f}')
        if valor > 0:
            litros_para_abastecer = valor / self._valor_litro
            self._abastecer(valor, litros_para_abastecer)
        else:
            print('O valor de ve ser maior que zero!')

    def abastecer_litro(self, litros_para_abastecer):
        print(f'\nAbastecendo {self.tipo_combustivel} por litro: {litros_para_abastecer:.2f}L')
        if litros_para_abastecer > 0:
            valor = litros_para_abastecer * self._valor_litro
            self._abastecer(valor, litros_para_abastecer)
        else:
            print('A quantidade de litros para abastecer de ve ser maior que zero!')

    def _abastecer(self, valo_pago, litros_abastecido):
        if litros_abastecido > self.quantidade_combustivel:
            print('Quantidade de combustível insuficiente na bomba!')
        else:
            self.quantidade_combustivel -= litros_abastecido
            print(f'''Valor Pago: R${valo_pago:.2f}
Litros Abastecidos: {litros_abastecido:.2f}L
Quantidade Combustível: {self._quantidade_combustivel:.2f}L''')


if __name__ == '__main__':
    bomb = BombaCombustivel('GNV', 6.54, 1000)
    bomb.abastecer_por_valor(1000)
    bomb.abastecer_litro(152.91)
    bomb.abastecer_por_valor(50)




Abastecendo GNV por valor: R$1000.00
Valor Pago: R$1000.00
Litros Abastecidos: 152.91L
Quantidade Combustível: 847.09L

Abastecendo GNV por litro: 152.91L
Valor Pago: R$1000.03
Litros Abastecidos: 152.91L
Quantidade Combustível: 694.18L

Abastecendo GNV por valor: R$50.00
Valor Pago: R$50.00
Litros Abastecidos: 7.65L
Quantidade Combustível: 686.54L
